contributors/design-proposals/pod-preemption.md

```diff
@@ -0,0 +1,187 @@
+
+## Terminology
+
+When a new pod has certain scheduling requirements that makes it infeasible on any node in the cluster, scheduler may choose to kill lower priority pods to satisfy the scheduling requirements of the new pod. We call this operation "**preemption**". Preemption is distinguished from "[**eviction**](https://kubernetes.io/docs/tasks/administer-cluster/out-of-resource/#eviction-policy)" where kubelet kills a pod on a node because that particular node is running out of resources.
+
+# Overview
+
+This document describes how preemption in Kubernetes works. Preemption is the action taken when an important pod requires resources or conditions which are not available in the cluster. So, one or more pods need to be killed to make room for the important pod. 
+
+# Detailed Design
+
+## Preemption scenario
+
+In this proposal, the only scenario under which a group of pods in Kubernetes may be preempted is when a higher priority pod cannot be scheduled due to various unmet scheduling requirements, such as lack of resources, unsatisfied affinity or anti-affinity rules, etc., and the preemption of the lower priority pods allows the higher priority pod to be scheduled. So, if the preemption of the lower priority pods does not help with scheduling of the higher priority pod, those lower priority pods will keep running and the higher priority pod will stay pending.  
+Please note the terminology here. The above scenario does not include "evictions" that are performed by the Kubelet when a node runs out of resources.  
+Please also note that scheduler may preempt a pod on one node in order to meet the scheduling requirements of a pending pod on another node. For example, if there is a low-priority pod running on node N in rack R, and there is a high-priority pending pod, and one or both of the pods have a requiredDuringScheduling anti-affinity rule saying they can't run on the same rack, then the lower-priority pod might be preempted to enable the higher-priority pod to schedule onto some node M != N on rack R (or, of course, M == N, which is the more standard same-node preemption scenario).
+
+## Scheduler performs preemption
+
+We propose preemption to be done by the scheduler -- it does it by deleting the being preempted pods. The component that performs the preemption must have the logic to find the right nodes for the pending pod. It must also have the logic to check whether preempting the chosen pods allows scheduling of the pending pod. These require the component to have the knowledge of predicate and priority functions.  
```

Comment:
I think I kind of assumed that the preemption will be implemented as extra service and the scheduler flow will be extended to handle not enough resources situations via it. That way it would not depend on any specific scheduler and so my comment would make sense (a way to  cooperate would be needed).